From: Steffen Maier <maier@linux.ibm.com>
Subject: scsi: zfcp: fix missing REC trigger trace on terminate_rport_io for ERP_FAILED
Patch-mainline: v4.18-rc1
Git-commit: d70aab55924b44f213fec2b900b095430b33eec6
References: bnc#1102087, LTC#168765

Description:  zfcp: fix tracing regressions, part 3
Symptom:      Cannot determine result of eh_host_reset_handler.

              Cannot see if and when zfcp retries abort or task management
              functions (LUN/target reset) while synchronizing scsi_eh and
              zfcp-internal recovery.

              Confusing zfcp ERP trace record if a SCSI device is deleted
              during scsi_eh host reset.

              Cannot determine that our terminate_rport_io callback was
              invoked if it returned early.

              Cannot determine that our terminate_rport_io callback was
              invoked if a long fast_io_fail_tmo was configured (typically
              longer than 12 seconds, e.g. 27 seconds).

              Cannot determine if something wanted to enqueue zfcp recovery
              for an object (adapter/port regular/unit) in ERP_FAILED state.

              Cannot determine if something wanted to enqueue zfcp recovery
              but no zfcp ERP thread was running.
Problem:      v2.6.35 commit a1dbfddd02d2 ("[SCSI] zfcp: Pass return code
              from fc_block_scsi_eh to scsi eh") introduced the first
              return with something other than the previously hardcoded
              single SUCCESS return path.

              Due to zfcp_erp_wait() and fc_block_scsi_eh() time can pass
              between the start of our eh callback and an actual send/recv
              of an abort / TMF request.

              If a SCSI device is deleted during scsi_eh host reset, we
              cannot get a reference to the SCSI device anymore since
              scsi_device_get returns !=0 by design. Assuming the recovery
              of adapter and port(s) was successful,
              zfcp_erp_strategy_followup_success() attempts to trigger a
              LUN reset for the half-gone SCSI device. However,
              zfcp_erp_setup_act() returns NULL as it cannot get the
              reference. Hence, zfcp_erp_action_enqueue() takes an early
              goto out and _NO_ recovery actually happens, but it writes a
              confusing trace record which states that zfcp will do a LUN
              recovery as "ERP need" is ZFCP_ERP_ACTION_REOPEN_LUN == 1 and
              equals "ERP want".
              Before v2.6.38 commit ae0904f60fab ("[SCSI] zfcp: Redesign of
              the debug tracing for recovery actions.") we could detect
              this case because the "erp_action" field in the trace was
              NULL. The rework removed erp_action as argument and field
              from the trace.

              If we get an fc_rport in terminate_rport_io() for which we
              cannot find a match within zfcp_get_port_by_wwpn(), the
              latter can return NULL. v2.6.30 commit 70932935b61e ("[SCSI]
              zfcp: Fix oops when port disappears") introduced an early
              return without adding a trace record for this case.

              After target cable pull: The ADISC for "link" test times out
              after typically 4 seconds. The open port recovery times out
              after typically 8 seconds, causing zfcp_port.status to
              contain ERP_FAILED. zfcp_scsi_terminate_rport_io() tries to
              enqueue forced port recovery, but due to ERP_FAILED this
              resulted in a silent early return.

              Trying to enqueue recovery for adapter / port regular / LUN
              resulted in silent early returns.

              Trying to enqueue recovery if no ERP thread was running
              resulted in silent early returns.
Solution:     Write trace record in zfcp area SCSI and re-use SCSI result
              field to contain the return value of eh_host_reset_handler.

              Add a trace before calling the two blocking functions
              zfcp_erp_wait() and fc_block_scsi_eh().

              Introduce an eyecatcher "flag" to mark the "ERP need" as 'not
              needed' but still keep the information which erp_action type,
              that zfcp_erp_required_act() had decided upon, is needed.
              0xc_ is chosen to be visibly different from 0x0_ in "ERP
              want".

              Write trace record in zfcp area REC with "ERP need" none.

              Write trace record in zfcp area REC with "ERP need" 0xe0.

              zfcp_erp_action_enqueue() already had two early outs which
              re-used the one zfcp_dbf_rec_trig() call. All ERP trigger
              functions finally run through zfcp_erp_action_enqueue(). So
              move the special handling for ZFCP_STATUS_COMMON_ERP_FAILED
              into zfcp_erp_action_enqueue() and add another early out with
              new trace marker for pseudo ERP need in this case. This
              removes all early returns from all ERP trigger functions so
              we always end up at zfcp_dbf_rec_trig().

              Write trace record in zfcp area REC with "ERP need" 0xc0.
Reproduction: If NPIV is not enabled with subchannel, ensure that your
              subchannel is the only one active on the entire channel.
              Set eh_deadline to a short enough time interval, e.g. 5
              seconds (before FCP device online: scsi_mod.eh_deadline=5).
              Enable RSCN suppression on the SAN switch port beyond the
              first link, i.e. towards the storage target. Disable that
              switch port. Send one SCSI command to the single path SCSI
              disk of that port in the background (because it will block
              for a while) e.g. via "dd if=/dev/sd... of=/dev/null
              iflag=direct count=1 &". After <SCSI command timeout>
              seconds, the command runs into the timeout. Due to the short
              eh_deadline, the SCSI midlayer immediately performs SCSI host
              reset. After the corresponding adapter recovery, the new
              trace record "schrh_r" appears.

              Difficult to reproduce as we need to have a SCSI command time
              out once (abort) or twice (scsi_eh), and timely have to
              prevent sending of the abort or TMF. Examples to prevent
              sending is severe memory pressure, or zfcp-internal recovery
              of the zfcp unit (e.g. write 0 to failed sysfs attribute) or
              of the target port (depending on potentially escalated
              scsi_eh scope) so the zfcp object is blocked.

              Difficult to reproduce as we need to get into scsi_eh, fully
              escalate to host reset, and something has to timely delete a
              SCSI device of this Scsi_Host.

              Disable auto port REscan with zfcp.no_auto_port_rescan=1. Let
              always enabled zfcp auto port scan attach a target remote
              port without attaching LUNs to it (zfcp.allow_lun_scan=0 and
              no explicit zfcp unit_add). Use zfcp port_remove to remove
              such target port again. After fast_io_fail_tmo or
              dev_loss_tmo, the new trace record "sctrpin" can appear.

              If NPIV is not enabled with subchannel, ensure that your
              subchannel is the only one active on the entire channel.
              Configure a large enough fast_io_fail_tmo, e.g. 27 seconds.
              Pull cable at target side. Wait until fast_io_fail_tmo runs
              out and with the fix a zfcp trace "sctrpi1" with "ERP need"
              0xe0 must appear in the REC area. (for shared non-NPIV
              subchannels, only the first subchannel, that happens to
              complete the corresponding code path, ends up in this
              particular case)

              Difficult to reproduce as we need a "soft" recovery trigger
              that does not clear ERP_FAILED (as opposed to e.g. failed
              sysfs attribute).

              No known reproduction.

Upstream-Description:

              scsi: zfcp: fix missing REC trigger trace on terminate_rport_io for ERP_FAILED

              For problem determination we always want to see when we were invoked on the
              terminate_rport_io callback whether we perform something or not.

              Temporal event sequence of interest with a long fast_io_fail_tmo of 27 sec:

              loose remote port

              t   workqueue
              [s] zfcp_q_<dev>       IRQ                 zfcperp<dev>

              === ================== =================== ============================

                0                    recv RSCN
                                     q p.test_link_work
                  block rport
                   start fast_io_fail_tmo
                  send ADISC ELS
                4                    recv ADISC fail
                                     block zfcp_port
                                                         port forced reopen
                                                         send open port
               12                    recv open port fail
                                                         q p.gid_pn_work
                                                         zfcp_erp_wakeup
                                                         (zfcp_erp_wait would return)
                  GID_PN fail

              Before this point, we got a SCSI trace with tag "sctrpi1" on fast_io_fail,
              e.g. with the typical 5 sec setting.

                  port.status |= ERP_FAILED

              If fast_io_fail_tmo triggers after this point, we missed a SCSI trace.

                  workqueue
                  fc_dl_<host>
                  ==================
               27 fc_timeout_fail_rport_io
                  fc_terminate_rport_io
                  zfcp_scsi_terminate_rport_io
                  zfcp_erp_port_forced_reopen
                  _zfcp_erp_port_forced_reopen
                   if (port.status & ERP_FAILED)
                    return;

              Therefore, write a trace before above early return.

              Example trace record formatted with zfcpdbf from s390-tools:

              Timestamp      : ...
              Area           : REC
              Subarea        : 00
              Level          : 1
              Exception      : -
              CPU ID         : ..
              Caller         : 0x...
              Record ID      : 1                      ZFCP_DBF_REC_TRIG
              Tag            : sctrpi1                SCSI terminate rport I/O
              LUN            : 0xffffffffffffffff                     none (invalid)
              WWPN           : 0x<wwpn>
              D_ID           : 0x<n_port_id>
              Adapter status : 0x...
              Port status    : 0x...
              LUN status     : 0x00000000                             none (invalid)
              Ready count    : 0x...
              Running count  : 0x...
              ERP want       : 0x03                   ZFCP_ERP_ACTION_REOPEN_PORT_FORCED
              ERP need       : 0xe0                   ZFCP_ERP_ACTION_FAILED

              Signed-off-by: Steffen Maier <maier@linux.ibm.com>
              Cc: <stable@vger.kernel.org> #2.6.38+
              Reviewed-by: Benjamin Block <bblock@linux.ibm.com>
              Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>


Signed-off-by: Steffen Maier <maier@linux.ibm.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 drivers/s390/scsi/zfcp_erp.c |   13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

--- a/drivers/s390/scsi/zfcp_erp.c
+++ b/drivers/s390/scsi/zfcp_erp.c
@@ -41,9 +41,13 @@ enum zfcp_erp_steps {
  * @ZFCP_ERP_ACTION_REOPEN_PORT_FORCED: Forced port recovery.
  * @ZFCP_ERP_ACTION_REOPEN_ADAPTER: Adapter recovery.
  * @ZFCP_ERP_ACTION_NONE: Eyecatcher pseudo flag to bitwise or-combine with
- *			  either of the other enum values.
+ *			  either of the first four enum values.
  *			  Used to indicate that an ERP action could not be
  *			  set up despite a detected need for some recovery.
+ * @ZFCP_ERP_ACTION_FAILED: Eyecatcher pseudo flag to bitwise or-combine with
+ *			    either of the first four enum values.
+ *			    Used to indicate that ERP not needed because
+ *			    the object has ZFCP_STATUS_COMMON_ERP_FAILED.
  */
 enum zfcp_erp_act_type {
 	ZFCP_ERP_ACTION_REOPEN_LUN         = 1,
@@ -51,6 +55,7 @@ enum zfcp_erp_act_type {
 	ZFCP_ERP_ACTION_REOPEN_PORT_FORCED = 3,
 	ZFCP_ERP_ACTION_REOPEN_ADAPTER     = 4,
 	ZFCP_ERP_ACTION_NONE		   = 0xc0,
+	ZFCP_ERP_ACTION_FAILED		   = 0xe0,
 };
 
 enum zfcp_erp_act_state {
@@ -378,8 +383,12 @@ static void _zfcp_erp_port_forced_reopen
 	zfcp_erp_port_block(port, clear);
 	zfcp_scsi_schedule_rport_block(port);
 
-	if (atomic_read(&port->status) & ZFCP_STATUS_COMMON_ERP_FAILED)
+	if (atomic_read(&port->status) & ZFCP_STATUS_COMMON_ERP_FAILED) {
+		zfcp_dbf_rec_trig(id, port->adapter, port, NULL,
+				  ZFCP_ERP_ACTION_REOPEN_PORT_FORCED,
+				  ZFCP_ERP_ACTION_FAILED);
 		return;
+	}
 
 	zfcp_erp_action_enqueue(ZFCP_ERP_ACTION_REOPEN_PORT_FORCED,
 				port->adapter, port, NULL, id, 0);
